src/Simplify_Let.cpp

```diff
@@ -195,14 +221,28 @@ Body Simplify::simplify_let(const LetOrLetStmt *op, ExprInfo *bounds) {
         VarInfo info = var_info.get(it->op->name);
         var_info.pop(it->op->name);
 
-        if (it->new_value.defined() && info.new_uses > 0) {
```

Comment:
You need to count loads or stores to "foo" as a usage of the name "foo" in the visitor above. That will make the hack below unnecessary.